SignalR Client to server streaming (#11126)
aspnet/AspNetCore.Docs: aspnetcore/signalr/streaming.md

By [Brennan Conroy](https://github.com/BrennanConroy)

::: moniker range="< aspnetcore-3.0"

ASP.NET Core SignalR supports streaming return values of server methods. This is useful for scenarios where fragments of data will come in over time. When a return value is streamed to the client, each fragment is sent to the client as soon as it becomes available, rather than waiting for all the data to become available.

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

> ASP.NET Core SignalR supports streaming from client to server and from server to client. This is useful for scenarios where fragments of data will come in over time. When streaming, each fragment is sent to the client or server as soon as it becomes available, rather than waiting for all the data to become available.

Reviewer: Update the `description: ` to indicate this article now also discusses client-to-server streaming. Feel free to rewrite/shorten it.

::: moniker-end

::: moniker range="= aspnetcore-2.1"

[View or download sample code](https://github.com/aspnet/Docs/tree/live/aspnetcore/signalr/streaming/sample.netcoreapp2.1) ([how to download](xref:index#how-to-download-a-sample))

::: moniker-end

::: moniker range="= aspnetcore-2.2"

[View or download sample code](https://github.com/aspnet/Docs/tree/live/aspnetcore/signalr/streaming/sample.netcoreapp2.2) ([how to download](xref:index#how-to-download-a-sample))

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

[View or download sample code](https://github.com/aspnet/Docs/tree/live/aspnetcore/signalr/streaming/sample.netcoreapp3.0) ([how to download](xref:index#how-to-download-a-sample))

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

A hub method automatically becomes a streaming hub method when it returns a `ChannelReader<T>`, `IAsyncEnumerable<T>`, `Task<ChannelReader<T>>`, or `Task<IAsyncEnumerable<T>>`.

::: moniker-end

::: moniker range="< aspnetcore-3.0"

A hub method automatically becomes a streaming hub method when it returns a `ChannelReader<T>` or a `Task<ChannelReader<T>>`.

::: moniker-end

> ## Set up the hub for streaming

Reviewer: Nit:

```suggestion
## Set up a hub for streaming
```

### Server to client streaming

> The following sample shows the basics of streaming data to the client using Channels. Whenever an object is written to the `ChannelWriter` that object is immediately sent to the client. At the end, the `ChannelWriter` is completed to tell the client the stream is closed.

Reviewer: We use ``<xref:System.Threading.Channels.ChannelReader`1>`` below so we should probably also reference ``<xref:System.Threading.Channels.ChannelWriter`1>`` here.

> [!NOTE]
> * Write to the `ChannelWriter` on a background thread and return the `ChannelReader` as soon as possible. Other hub invocations will be blocked until a `ChannelReader` is returned.
> * Wrap your logic in a `try ... catch` and complete the `Channel` in the catch and outside the catch to make sure the hub method invocation is completed properly.

::: moniker range="= aspnetcore-2.1"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.1/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range="= aspnetcore-2.2"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.2/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp3.0/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range=">= aspnetcore-2.2"

> In ASP.NET Core 2.2 or later, server to client streaming Hub methods can accept a `CancellationToken` parameter that will be triggered when the client unsubscribes from the stream. Use this token to stop the server operation and release any resources if the client disconnects before the end of the stream.

Reviewer 2: I'll regret asking this but why not. Adding @scottaddie here to get his 411 on it. Should we say "server-to-client" and "client-to-server" instead of the non-hyphenated versions?

Reviewer: 👍 for the hyphenation. In this case we have "server to client streaming Hub methods" which really is three adjectives describing the "methods":

1. server-to-client
2. streaming
3. hub

The hyphenation makes this more clear.

Also nit: It looks like in the rest of the docs we don't treat "hub" as a proper noun, so "Hub" in "Hub method" should be lowercase.

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Client to server streaming

> A hub method automatically becomes a client to server streaming hub method when it accepts a `ChannelReader<T>`. Below is a sample that shows the basics of reading streaming data from the client. Whenever the client writes to the stream the data is written into the `ChannelReader` on the server which the hub method should be reading from.

Reviewer 2: Should we link to the ChannelReader article or reference page?

Reviewer 3: We should xref to the reference page in at least some of these so the user can get there. Usually we don't code-fence when xreffing, so it would be `accepts a [ChannelReader<T>](xref:...). Below...`

Author: How do xref to dotnet section of docs?!

Reviewer 3: Paging @scottaddie, @tdykstra ? :)

Reviewer 4: @guardrex

Reviewer 5: https://xrefgen.azurewebsites.net/

Reviewer 6: ```
<xref:System.Threading.Channels.ChannelReader`1>
```

... if you're feel'in lucky 🍀, you can try this thing: https://xrefgen.azurewebsites.net/

... but only if your PC is paid for and a **_fried MOBO_** is something u can deal with. 😀 

i.e., that thing isn't 100% accurate. Look at the build report to see if the link doesn't work. If it isn't correct, pull up the API object/member in the API Browser, go into the `<meta>` tags and look for `ms.assetid` ... it's value should work. 🤞

Reviewer 5: I don't think  types `<T>` work. For those use the `[](/dotnet/api/aspnetcore)` link.

@guardrex 	will correct me if I'm wrong.

Reviewer 6: I just made updates to the XREF gen thing to compensate for `<T>`.

That app is a WIP, but it should work in *most* cases.

The ultimate source of truth is pulling up the API Browser page and looking at the `ms.assetid` meta tag value.

Absolutely *Yes!* on `/dotnet/api` links. If that's easier for engineers do that. We'll update them on a future doc update. IMO, do whatever is *fastest* 🏃.

Author: Worked

By [Brennan Conroy](https://github.com/BrennanConroy)

::: moniker range="< aspnetcore-3.0"

ASP.NET Core SignalR supports streaming return values of server methods. This is useful for scenarios where fragments of data will come in over time. When a return value is streamed to the client, each fragment is sent to the client as soon as it becomes available, rather than waiting for all the data to become available.

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

ASP.NET Core SignalR supports streaming from client-to-server and from server-to-client. This is useful for scenarios where fragments of data will come in over time. When streaming, each fragment is sent to the client or server as soon as it becomes available, rather than waiting for all the data to become available.

::: moniker-end

> ::: moniker range="= aspnetcore-2.1"

Reviewer 7: The doc will be easier to maintain if you use the following link and display it for all versions:
https://github.com/aspnet/Docs/tree/live/aspnetcore/signalr/streaming

Author: I don't know what you mean. The sample has changed every release, shouldn't the user be able to see the correct sample based on what moniker they are looking at?

> * Wrap your logic in a `try ... catch` and complete the `Channel` in the catch and outside the catch to make sure the hub method invocation is completed properly.

::: moniker range="= aspnetcore-2.1"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.1/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range="= aspnetcore-2.2"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.2/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp3.0/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range=">= aspnetcore-2.2"

In ASP.NET Core 2.2 or later, server-to-client streaming hub methods can accept a `CancellationToken` parameter that will be triggered when the client unsubscribes from the stream. Use this token to stop the server operation and release any resources if the client disconnects before the end of the stream.

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Client to server streaming

> A hub method automatically becomes a client-to-server streaming hub method when it accepts a <xref:System.Threading.Channels.ChannelReader`1>. The following sample shows the basics of reading streaming data from the client. Whenever the client writes to the <xref:System.Threading.Channels.ChannelWriter`1>, the data is written into the `ChannelReader` on the server which the hub method should be reading from.

Reviewer: nit:

```suggestion
A hub method automatically becomes a client-to-server streaming hub method when it accepts a <xref:System.Threading.Channels.ChannelReader`1>. The following sample shows the basics of reading streaming data sent from the client. Whenever the client writes to the <xref:System.Threading.Channels.ChannelWriter`1>, the data is written into the `ChannelReader` on the server which the hub method should be reading from.
```

For extra clarity. Otherwise it could be interpreted as "reading data using the client".

> A hub method automatically becomes a client to server streaming hub method when it accepts a <xref:System.Threading.Channels.ChannelReader`1>. Below is a sample that shows the basics of reading streaming data from the client. Whenever the client writes to the <xref:System.Threading.Channels.ChannelWriter`1> the data is written into the `ChannelReader` on the server which the hub method should be reading from.

Reviewer 7: ```suggestion
A hub method automatically becomes a client-to-server streaming hub method when it accepts a <xref:System.Threading.Channels.ChannelReader`1>. The following sample shows the basics of reading streaming data from the client. Whenever the client writes to the <xref:System.Threading.Channels.ChannelWriter`1>, the data is written into the `ChannelReader` on the server which the hub method should be reading from.
```

[!code-csharp[Streaming upload hub method](streaming/sample.netcoreapp3.0/Hubs/StreamHub.cs?name=snippet2)]

::: moniker-end

## .NET client

### Server to client streaming

> The `StreamAsChannelAsync` method on `HubConnection` is used to invoke a server-to-client streaming method. Pass the hub method name, and arguments defined in the hub method to `StreamAsChannelAsync`. The generic parameter on `StreamAsChannelAsync<T>` specifies the type of objects returned by the streaming method. A `ChannelReader<T>` is returned from the stream invocation, and represents the stream on the client. To read data, a common pattern is to loop over `WaitToReadAsync` and call `TryRead` when data is available. The loop will end when the stream has been closed by the server, or the cancellation token passed to `StreamAsChannelAsync` is canceled.

Reviewer 7: ```suggestion
The `StreamAsChannelAsync` method on `HubConnection` is used to invoke a server-to-client streaming method. Pass the hub method name, and arguments defined in the hub method, to `StreamAsChannelAsync`. The generic parameter on `StreamAsChannelAsync<T>` specifies the type of objects returned by the streaming method. A `ChannelReader<T>` is returned from the stream invocation, and represents the stream on the client. To read data, a common pattern is to loop over `WaitToReadAsync` and call `TryRead` when data is available. The loop will end when the stream has been closed by the server, or the cancellation token passed to `StreamAsChannelAsync` is canceled.
```

> The `StreamAsChannelAsync` method on `HubConnection` is used to invoke a server-to-client streaming method. Pass the hub method name, and arguments defined in the hub method, to `StreamAsChannelAsync`. The generic parameter on `StreamAsChannelAsync<T>` specifies the type of objects returned by the streaming method. A `ChannelReader<T>` is returned from the stream invocation, and represents the stream on the client. To read data, a common pattern is to loop over `WaitToReadAsync` and call `TryRead` when data is available. The loop will end when the stream has been closed by the server, or the cancellation token passed to `StreamAsChannelAsync` is canceled.

Reviewer: I don't think it's worth calling out the WaitToReadAsync and TryRead ChannelReader methods in the prose. Leave the channel API usage to the samples like you did for the ChannelWriter methods.

Reviewer 2: agreed, and if folks need more we can add more later

::: moniker range="= aspnetcore-2.1"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.1/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range="= aspnetcore-2.2"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.2/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp3.0/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range=">= aspnetcore-2.2"

In ASP.NET Core 2.2 or later, server to client streaming hub methods can accept a `CancellationToken` parameter that will be triggered when the client unsubscribes from the stream. Use this token to stop the server operation and release any resources if the client disconnects before the end of the stream.

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

> In ASP.NET Core 3.0 or later, streaming hub methods can return `IAsyncEnumerable<T>` in addition to `ChannelReader<T>`. The simplest way to return `IAsyncEnumerable<T>` is by making the hub method an async iterator method as the following sample demonstrates. Hub async iterator methods can accept a `CancellationToken` parameter that will be triggered when the client unsubscribes from the stream. Async iterator methods easily avoid problems common with Channels such as not returning the `ChannelReader` early enough or exiting the method without completing the `ChannelWriter`.

Reviewer: What do you think of moving this before the Channels section since async iterator methods are the more convenient way to generate streams?

Author: I'm fine with it

[!INCLUDE[](~/includes/csharp-8-required.md)]

[!code-csharp[Streaming hub async iterator method](streaming/sample.netcoreapp3.0/Hubs/AsyncEnumerableHub.cs?name=snippet_AsyncIterator)]

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Client to server streaming

> A hub method automatically becomes a client to server streaming hub method when it accepts a <xref:System.Threading.Channels.ChannelReader`1>. Below is a sample that shows the basics of reading streaming data from the client. Whenever the client writes to the stream the data is written into the `ChannelReader` on the server which the hub method should be reading from.

Reviewer: ```suggestion
A hub method automatically becomes a client to server streaming hub method when it accepts a <xref:System.Threading.Channels.ChannelReader`1>. The following sample shows the basics of reading streaming data from the client. Whenever the client writes to the stream, the data is made available via the `ChannelReader` on the server which the hub method should be reading from.
```

Reviewer: You probably also want to point out that you can only have one `ChannelReader<T>` param per method unless you can actually have more.

Author: You can have more than 1

Reviewer: Also since this is the server-side docs, we should probably include a reference to ChannelWriter. Maybe replace "Whenever the client writes to the stream" with "Whenever the client writes to the`<xref:System.Threading.Channels.ChannelWriter`1>`".

[!code-csharp[Streaming upload hub method](streaming/sample.netcoreapp3.0/Hubs/StreamHub.cs?name=snippet2)]

::: moniker-end

## .NET client

### Server to client streaming

> The `StreamAsChannelAsync` method on `HubConnection` is used to invoke a server to client streaming method. Pass the hub method name, and arguments defined in the hub method to `StreamAsChannelAsync`. The generic parameter on `StreamAsChannelAsync<T>` specifies the type of objects returned by the streaming method. A `ChannelReader<T>` is returned from the stream invocation, and represents the stream on the client. To read data, a common pattern is to loop over `WaitToReadAsync` and call `TryRead` when data is available. The loop will end when the stream has been closed by the server, or the cancellation token passed to `StreamAsChannelAsync` is canceled.

Reviewer: I like using xref's for the ChannelReader/Writer refs in the client docs too

Author: Do we want to xref every single mention, or just the first time on every page?

Reviewer: I think first time is OK for the most part. If you wouldn't be able to see the xref when looking at one of the monikers like aspnetcore-2.2, I would say duplicate the xref to make sure the link is visible in all views.

::: moniker range="= aspnetcore-2.1"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.1/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range="= aspnetcore-2.2"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.2/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

[!code-csharp[Streaming hub method](streaming/sample.netcoreapp3.0/Hubs/StreamHub.cs?name=snippet1)]

::: moniker-end

::: moniker range=">= aspnetcore-2.2"

In ASP.NET Core 2.2 or later, server to client streaming hub methods can accept a `CancellationToken` parameter that will be triggered when the client unsubscribes from the stream. Use this token to stop the server operation and release any resources if the client disconnects before the end of the stream.

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Client to server streaming

> A hub method automatically becomes a client to server streaming hub method when it accepts a `ChannelReader<T>`. Below is a sample that shows the basics of reading streaming data from the client. Whenever the client writes to the stream the data is written into the `ChannelReader` on the server which the hub method should be reading from.

Reviewer 3: Do we accept `IAsyncEnumerable<T>` on the server?

Author: Not yet https://github.com/aspnet/AspNetCore/issues/7930

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Client to server streaming

> To invoke a client-to-server streaming hub method from the .NET Client create a `Channel` and pass the `ChannelReader` as an argument to `SendAsync`, `InvokeAsync`, or `StreamAsChannelAsync`, depending on the hub method being invoked.

Reviewer 7: ```suggestion
To invoke a client-to-server streaming hub method from the .NET client create a `Channel` and pass the `ChannelReader` as an argument to `SendAsync`, `InvokeAsync`, or `StreamAsChannelAsync`, depending on the hub method being invoked.
```

Whenever data is written to the `ChannelWriter` the hub method on the server will receive a new item with the data from the client.

> To end the stream, complete the channel with `channel.Writer.TryComplete()`.

Reviewer: Why recommend TryComplete() over Complete() here?

Reviewer 3: Same reason we generally use `TrySetResult` and friends. `Complete` will throw if the channel is already completed.

Reviewer: I prefer having the Channel throw in simple cases like these. No other code has access to the Writer, so double completion would never be expected and would always be a bug.

Reviewer: @BrennanConroy You asked for feedback. I stand by my last comment here.

Reviewer: If we weren't meant to use `ChannelWriter.Complete()` in cases like these, the method wouldn't exist.

Author: I thought I changed this, must have forgotten.

```csharp
var channel = Channel.CreateBounded<string>(10);
await connection.SendAsync("UploadStream", channel.Reader);
await channel.Writer.WriteAsync("some data");
await channel.Writer.WriteAsync("some more data");

> await channel.Writer.TryComplete();

Reviewer: I don't think you can await TryComplete()

* Arguments defined in the hub method. In the following example, the arguments are: a count for the number of stream items to receive, and the delay between stream items.

`connection.stream` returns an `IStreamResult` which contains a `subscribe` method. Pass an `IStreamSubscriber` to `subscribe` and set the `next`, `error`, and `complete` callbacks to get notifications from the `stream` invocation.

::: moniker range="= aspnetcore-2.1"

[!code-javascript[Streaming javascript](streaming/sample.netcoreapp2.1/wwwroot/js/stream.js?range=19-36)]

To end the stream from the client, call the `dispose` method on the `ISubscription` that is returned from the `subscribe` method.

::: moniker-end

::: moniker range=">= aspnetcore-2.2"

[!code-javascript[Streaming javascript](streaming/sample.netcoreapp2.2/wwwroot/js/stream.js?range=19-36)]

To end the stream from the client, call the `dispose` method on the `ISubscription` that is returned from the `subscribe` method. Calling this method will cause the `CancellationToken` parameter of the Hub method (if you provided one) to be canceled.

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Client to server streaming

> JavaScript clients call client to server streaming methods on hubs by passing in a `Subject` as one of the arguments in `send`, `invoke`, or `stream`, depending on the hub method being invoked. The `Subject` needs to be a class that looks like a `Subject`. For example, if you use RxJS then you can use the `Subject` class from that library.

Reviewer: Can we link to some external docs surrounding `Subject`?

Reviewer 3: We should have API docs (eventually). As long as there are TSDoc comments in place. We can also link to RxJS docs on Subject.

* Arguments defined in the hub method. In the following example, the arguments are: a count for the number of stream items to receive, and the delay between stream items.

`connection.stream` returns an `IStreamResult` which contains a `subscribe` method. Pass an `IStreamSubscriber` to `subscribe` and set the `next`, `error`, and `complete` callbacks to get notifications from the `stream` invocation.

::: moniker range="= aspnetcore-2.1"

[!code-javascript[Streaming javascript](streaming/sample.netcoreapp2.1/wwwroot/js/stream.js?range=19-36)]

To end the stream from the client, call the `dispose` method on the `ISubscription` that is returned from the `subscribe` method.

::: moniker-end

::: moniker range=">= aspnetcore-2.2"

[!code-javascript[Streaming javascript](streaming/sample.netcoreapp2.2/wwwroot/js/stream.js?range=19-36)]

To end the stream from the client, call the `dispose` method on the `ISubscription` that is returned from the `subscribe` method. Calling this method will cause the `CancellationToken` parameter of the Hub method (if you provided one) to be canceled.

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Client to server streaming

> JavaScript clients call client-to-server streaming methods on hubs by passing in a `Subject` as one of the arguments in `send`, `invoke`, or `stream`, depending on the hub method being invoked. The `Subject` needs to be a class that looks like a `Subject`. For example, if you use RxJS then you can use the [Subject](https://rxjs-dev.firebaseapp.com/api/index/class/Subject) class from that library.

Reviewer 7: ```suggestion
JavaScript clients call client-to-server streaming methods on hubs by passing in a `Subject` as one of the arguments in `send`, `invoke`, or `stream`, depending on the hub method being invoked. The `Subject` needs to be a class that looks like a `Subject`. For example, in RxJS you can use the [Subject](https://rxjs-dev.firebaseapp.com/api/index/class/Subject) class from that library.
```

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Client to server streaming

> To invoke a client-to-server streaming hub method from the .NET client create a `Channel` and pass the `ChannelReader` as an argument to `SendAsync`, `InvokeAsync`, or `StreamAsChannelAsync`, depending on the hub method being invoked.

Reviewer: Should we use hyphens for client-to-server and server-to-client in our subsection headers like we do for our prose?

Reviewer 6: Yes (adjective)

Reviewer 2: IMO yes (but OCD)

* Arguments defined in the hub method. In the following example, the arguments are: a count for the number of stream items to receive, and the delay between stream items.

`connection.stream` returns an `IStreamResult` which contains a `subscribe` method. Pass an `IStreamSubscriber` to `subscribe` and set the `next`, `error`, and `complete` callbacks to get notifications from the `stream` invocation.

::: moniker range="= aspnetcore-2.1"

[!code-javascript[Streaming javascript](streaming/sample.netcoreapp2.1/wwwroot/js/stream.js?range=19-36)]

To end the stream from the client, call the `dispose` method on the `ISubscription` that is returned from the `subscribe` method.

::: moniker-end

::: moniker range=">= aspnetcore-2.2"

[!code-javascript[Streaming javascript](streaming/sample.netcoreapp2.2/wwwroot/js/stream.js?range=19-36)]

To end the stream from the client, call the `dispose` method on the `ISubscription` that is returned from the `subscribe` method. Calling this method will cause the `CancellationToken` parameter of the Hub method (if you provided one) to be canceled.

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Client to server streaming

> JavaScript clients call client-to-server streaming methods on hubs by passing in a `Subject` as one of the arguments in `send`, `invoke`, or `stream`, depending on the hub method being invoked. The `Subject` needs to be a class that looks like a `Subject`. For example, in RxJS you can use the [Subject](https://rxjs-dev.firebaseapp.com/api/index/class/Subject) class from that library.

Reviewer: You can have multiple streaming arguments and put the streaming arguments in any position (e.g. they don't have to be the first arguments), right?

Phrasing like "by passing in a `Subject` as one of the arguments" is making me doubt this.

In the three "Client to server streaming" subsections, we should make it clear that any number of streaming arguments in any position is supported. We should have a sample that takes streams as the 1st and 3rd arguments or something.

        () -> {/* Define your onCompleted handler here. */});
```

> The `stream` method on `HubConnection` returns an Observable of the stream item type. The Observable type's `subscribe` method is where `onNext`,  `onError` and  `onCompleted` handlers are defined.

Reviewer 7: ```suggestion
The `stream` method on `HubConnection` returns an Observable of the stream item type. The Observable type's `subscribe` method is where `onNext`, `onError` and `onCompleted` handlers are defined.
```
